core/amber/src/main/python/core/models/operator.py

```diff
@@ -49,8 +64,44 @@ def process_tuple(self, tuple_: Union[Tuple, InputExhausted], input_: int) -> It
         """
         pass
 
-    def close(self) -> None:
+
+class TableOperator(TupleOperator):
+    """
+    Base class for table-oriented operators. A concrete implementation must
+    be provided upon using.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.__internal_is_source: bool = False
+        self.__table_data: Mapping[int, List[Tuple]] = defaultdict(list)
+
+    @overrides.final
+    def process_tuple(self, tuple_: Union[Tuple, InputExhausted], input_: int) -> Iterator[Optional[TupleLike]]:
         """
-        Close the context of the operator.
+        Process an input Tuple from the given link. The Tuple is represented as pandas.Series.
+        :param tuple_: Union[Tuple, InputExhausted], either
+                        1. a Tuple from a link to be processed;
+                        2. an InputExhausted indicating no more data from this link.
+
+        :param input_: int, input index of the current Tuple.
+        :return: Iterator[Optional[TupleLike]], producing one TupleLike object at a time, or None.
+        """
+        if isinstance(tuple_, Tuple):
+            self.__table_data[input_].append(tuple_)
+        else:
+            table = Table(pandas.DataFrame(self.__table_data[input_]))
+            for output_table in self.process_table(table, input_):
+                for _, output_tuple in output_table.iterrows():
```

Comment:
maybe need to check for null for output_table

Reply:
done

Reply:
Wait... where's the null check?